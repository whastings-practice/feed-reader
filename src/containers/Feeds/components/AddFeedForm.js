import React from 'react';
import { withFormik } from 'formik';

import Button from 'reactstrap/lib/Button';
import FormGroup from 'reactstrap/lib/FormGroup';
import Label from 'reactstrap/lib/Label';
import Input from 'reactstrap/lib/Input';
import Modal from 'reactstrap/lib/Modal';
import ModalHeader from 'reactstrap/lib/ModalHeader';
import ModalBody from 'reactstrap/lib/ModalBody';
import ModalFooter from 'reactstrap/lib/ModalFooter';

function AddFeedForm(props) {
  const { handleChange, handleSubmit, values } = props;

  return (
    <Modal isOpen={props.isOpen} toggle={props.onClose}>
      <ModalHeader toggle={props.onClose}>Add Feed</ModalHeader>
      <form onSubmit={handleSubmit}>
        <ModalBody>
          <FormGroup>
            <Label htmlFor="feed-url-input">URL:</Label>
            <Input
              id="feed-url-input"
              name="url"
              type="url"
              onChange={handleChange}
              value={values.url}
              required
            />
          </FormGroup>
        </ModalBody>
        <ModalFooter>
          <Button color="primary">Add</Button>
          <Button color="Secondary" onClick={props.onClose} type="button">
            Cancel
          </Button>
        </ModalFooter>
      </form>
    </Modal>
  );
}

export default withFormik({
  mapPropsToValues: () => ({
    url: '',
  }),
  handleSubmit: (values, { props }) => props.onSubmit(values.url),
})(AddFeedForm);
