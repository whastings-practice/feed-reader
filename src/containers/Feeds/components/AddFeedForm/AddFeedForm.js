import React from 'react';
import { Formik } from 'formik';

import Alert from 'reactstrap/lib/Alert';
import Button from 'reactstrap/lib/Button';
import FormGroup from 'reactstrap/lib/FormGroup';
import Label from 'reactstrap/lib/Label';
import Input from 'reactstrap/lib/Input';
import Modal from 'reactstrap/lib/Modal';
import ModalHeader from 'reactstrap/lib/ModalHeader';
import ModalBody from 'reactstrap/lib/ModalBody';
import ModalFooter from 'reactstrap/lib/ModalFooter';

import Autofocus from '../../../../components/Autofocus';
import LoadingSpinner from '../../../../components/LoadingSpinner/LoadingSpinner';

function renderForm(props) {
  const { errors, handleChange, handleSubmit, onClose, isSubmitting, values } = props;

  const urlError = errors && errors.url;

  return (
    <form onSubmit={handleSubmit}>
      <ModalHeader toggle={onClose}>Add Feed</ModalHeader>
      <ModalBody>
        {errors &&
          <Autofocus>
            <Alert color="danger">
              Adding feed failed.
            </Alert>
          </Autofocus>
        }
        <FormGroup>
          <Label htmlFor="feed-url-input">URL:</Label>
          <Input
            id="feed-url-input"
            name="url"
            type="url"
            onChange={handleChange}
            value={values.url}
            disabled={isSubmitting}
            className={urlError ? 'is-invalid' : ''}
            aria-invalid={!!urlError}
            aria-describedby="feed-url-input-desc"
            required
          />
          <div id="feed-url-input-desc" className="invalid-feedback">
            {urlError}
          </div>
        </FormGroup>
      </ModalBody>
      <ModalFooter>
        {isSubmitting && <LoadingSpinner />}
        <Button color="primary" disabled={isSubmitting}>Add</Button>
        <Button
          className="qa-close-btn"
          color="Secondary"
          onClick={onClose}
          type="button"
          disabled={isSubmitting}
        >
          Cancel
        </Button>
      </ModalFooter>
    </form>
  );
}

export default function AddFeedForm(props) {
  if (!props.isOpen) {
    return null;
  }

  return (
    <Modal isOpen={true} toggle={props.onClose}>
      <Formik
        initialValues={{ url: '' }}
        onSubmit={(values) => props.onSubmit(values.url)}
        render={(fProps) => renderForm({ ...fProps, ...props })}
      />
    </Modal>
  );
}
