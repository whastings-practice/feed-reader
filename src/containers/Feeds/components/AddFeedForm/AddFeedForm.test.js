import React from 'react';
import { mount, shallow } from 'enzyme';

import AddFeedForm from './AddFeedForm';

const noop = () => {};

function CallRenderFunction({ fn }) {
  return fn();
}

describe('AddFeedForm', () => {
  let props;
  let renderOutput;
  let wrapper;

  function shallowRender() {
    wrapper = shallow(<AddFeedForm {...props} />);
    if (props.isOpen) {
      const renderForm = wrapper.find('Formik').prop('render');
      renderOutput = shallow(<CallRenderFunction fn={renderForm} />);
    }
  }

  function mountRender() {
    wrapper = mount(<AddFeedForm {...props} />);
  }

  beforeEach(() => {
    props = {
      handleChange: noop,
      handleSubmit: noop,
      isOpen: false,
      isSubmitting: false,
      onClose: noop,
      values: { url: '' },
    };
    shallowRender();
  });

  test('it renders nothing by default', () => {
    expect(wrapper.get(0)).toBeNull();
  });

  describe('when open', () => {
    beforeEach(() => {
      props.isOpen = true;
      props.onClose = jest.fn();
      shallowRender();
    });

    test('it renders a form inside a modal', () => {
      expect(wrapper).toMatchSnapshot();
      expect(renderOutput).toMatchSnapshot();
    });

    test('it calls onClose prop when modal triggers close', () => {
      wrapper.find('Modal').prop('toggle')();
      expect(props.onClose).toHaveBeenCalledTimes(1);

      renderOutput.find('ModalHeader').prop('toggle')();
      expect(props.onClose).toHaveBeenCalledTimes(2);
    });

    test('it calls onClose prop when close button is clicked', () => {
      renderOutput.find('.qa-close-btn').simulate('click');
      expect(props.onClose).toBeCalled();
    });

    describe('when isSubmitting prop is true', () => {
      beforeEach(() => {
        props.isSubmitting = true;
        shallowRender();
      });

      test('it disables the text input and the buttons', () => {
        expect(renderOutput.find('Input')).toHaveProp('disabled', true);
        renderOutput.find('Button').forEach((buttonWrapper) => {
          expect(buttonWrapper).toHaveProp('disabled', true);
        });
      });

      test('it displays a loading spinner', () => {
        expect(renderOutput.find('LoadingSpinner')).toBePresent();
      });
    });

    describe('with errors', () => {
      beforeEach(() => {
        props.errors = { url: 'URL is invalid' };
        shallowRender();
      });

      test('it renders an autofocused alert', () => {
        const alert = renderOutput.find('Alert');
        expect(alert).toHaveProp('color', 'danger');
        expect(alert.parent()).toMatchSelector('Autofocus');
      });

      test('it marks input as invalid and associates an error message', () => {
        const input = renderOutput.find('Input');
        const errorMessage = renderOutput.find(`#${input.prop('aria-describedby')}`);
        expect(input).toHaveClassName('is-invalid');
        expect(input).toHaveProp('aria-invalid', true);
        expect(errorMessage).toHaveText(props.errors.url);
      });
    });
  });
});
