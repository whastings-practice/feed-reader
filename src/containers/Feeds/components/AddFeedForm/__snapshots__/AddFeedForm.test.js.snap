// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddFeedForm when open it renders a form inside a modal 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransitionTimeout={150}
  fade={true}
  isOpen={true}
  keyboard={true}
  modalTransitionTimeout={300}
  toggle={[Function]}
  zIndex={1050}
>
  <Formik
    enableReinitialize={false}
    initialValues={
      Object {
        "url": "",
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    render={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
  />
</Modal>
`;

exports[`AddFeedForm when open it renders a form inside a modal 2`] = `
<form
  onSubmit={[Function]}
>
  <ModalHeader
    tag="h4"
    toggle={[Function]}
    wrapTag="div"
  >
    Add Feed
  </ModalHeader>
  <ModalBody
    tag="div"
  >
    <FormGroup
      tag="div"
    >
      <Label
        htmlFor="feed-url-input"
        tag="label"
      >
        URL:
      </Label>
      <Input
        aria-describedby="feed-url-input-desc"
        aria-invalid={false}
        className=""
        disabled={false}
        id="feed-url-input"
        name="url"
        onChange={[Function]}
        required={true}
        tag="p"
        type="url"
        value=""
      />
      <div
        className="invalid-feedback"
        id="feed-url-input-desc"
      />
    </FormGroup>
  </ModalBody>
  <ModalFooter
    tag="div"
  >
    <Button
      color="primary"
      disabled={false}
      tag="button"
    >
      Add
    </Button>
    <Button
      className="qa-close-btn"
      color="Secondary"
      disabled={false}
      onClick={[Function]}
      tag="button"
      type="button"
    >
      Cancel
    </Button>
  </ModalFooter>
</form>
`;
